### ISSUE TITLE
[Bug]: Moving the cursor keeps the previous suggestion open

### ISSUE THREAD

**Issue by MrMage** [intitni/CopilotForXcode#197]
### Before Reporting

- [X] I have checked FAQ, and there is no solution to my issue
- [X] I have searched the existing issues, and there is no existing issue for my issue

### What happened?

When I type into the window, the suggestion appears. But when I then move the cursor to another line, the window follows the cursor — with the same suggestion as before. I would have expected the window to close (until I start typing again) when moving the cursor, because the old suggestions is no longer relevant.

Please see the following video for a demonstration:

https://user-images.githubusercontent.com/117466/236443234-95cbc814-bf24-4f5a-9e26-e3367d1ce2fe.mp4

Note that I have the "real-time suggestion" open enabled; debounce delay is set to 0.7 s.

### How to reproduce the bug.

1. Move the cursor to some location.
2. Start typing.
3. The suggestion window appears. So far, so good.
4. Move the cursor elsewhere, e.g. by clicking.
5. The suggestion window moves to the new line, but with the old suggestion.

### macOS version

13.3.1 (a) (with RSR)

### Xcode version

14.3

### Copilot for Xcode version

0.14.1

**Comment by MrMage** [comment]
With a lot of help from GPT-4, I arrived at the following to observe the selected text of the focused UI element. When the selected text range changes, this also triggers, even when the selected text is "" both before and after "" (e.g. when merely moving the cursor).

This is very far from production code, but it seems to work. For example, this currently fixates on a single focused UI element; i.e. it only accounts for a single editor. In reality, one would also need to observe the focused UI element, and whenever that changes, deregister the old observer and register a new one for the newly-focused element.

Maybe it helps :-)

```swift
func _toAXUIElement(_ element: Any?) -> AXUIElement? {
	guard let element = element else { return nil }
	return (element as! AXUIElement)
}

@objc public class SelectedTextMonitor: NSObject {
	private let appElement: NSRunningApplication
	private var observedUIElement: AXUIElement
	private var observer: AXObserver?  // Must be retained while we want to observe

	@objc public init?(bundleIdentifier: String) {
		guard let app = NSRunningApplication.runningApplications(withBundleIdentifier: bundleIdentifier).first else {
			print("Application not found.")
			return nil
		}

		self.appElement = app
		self.observedUIElement = AXUIElementCreateApplication(app.processIdentifier)
		super.init()

		// Let's get the text editor element using AXUIElementCopyAttributeValue instead:
		var result: AnyObject?
		AXUIElementCopyAttributeValue(observedUIElement, kAXFocusedUIElementAttribute as CFString, &result)
		print("observedUIElement:", result)

		observedUIElement = _toAXUIElement(result)!

		// And let's get the selected range using AXUIElementCopyAttributeValue instead:
		print("selectedTextRange:", getSelectedTextRange())
	}

	private func getSelectedTextRange() -> AnyObject? {
		var selectedTextRange: AnyObject?
		AXUIElementCopyAttributeValue(observedUIElement, kAXSelectedTextRangeAttribute as CFString, &selectedTextRange)
		return selectedTextRange
	}

	@objc public func startMonitoringSelectedText() {
		AXObserverCreate(appElement.processIdentifier, { (observer, observedElement, notificationName, context) in
			let monitor = Unmanaged<SelectedTextMonitor>.fromOpaque(context!).takeUnretainedValue()
			guard let selectedText = monitor.getSelectedText(from: observedElement) else {
				return
			}

			print("Selected text: \(selectedText)")
			print("Selected text range: \(monitor.getSelectedTextRange())")
		}, &observer)

		guard let axObserver = observer else {
			print("Failed to create AXObserver.")
			return
		}

		let selfRef = Unmanaged.passUnretained(self).toOpaque()
		AXObserverAddNotification(axObserver, observedUIElement, kAXSelectedTextChangedNotification as CFString, selfRef)
		CFRunLoopAddSource(CFRunLoopGetCurrent(), AXObserverGetRunLoopSource(axObserver), .defaultMode)
	}

	private func getSelectedText(from element: AXUIElement) -> String? {
		var result: AnyObject?
		let status = AXUIElementCopyAttributeValue(element, kAXSelectedTextAttribute as CFString, &result)

		guard status == .success, let selectedText = result as? String else {
			return nil
		}

		return selectedText
	}
}
```

Sample output:

```
observedUIElement: Optional(<AXUIElement 0x600003155530> {pid=41724})
selectedTextRange: Optional(<AXValue 0x600003155440> {value = location:14990 length:0 type = kAXValueCFRangeType})
Selected text: 
Selected text range: Optional(<AXValue 0x600003144000> {value = location:15031 length:0 type = kAXValueCFRangeType})
Selected text: 
Selected text range: Optional(<AXValue 0x6000031659b0> {value = location:14948 length:0 type = kAXValueCFRangeType})
Selected text: 
Selected text range: Optional(<AXValue 0x600003164030> {value = location:15032 length:0 type = kAXValueCFRangeType})
Selected text: 
Selected text range: Optional(<AXValue 0x600003144090> {value = location:14997 length:0 type = kAXValueCFRangeType})
Selected text: getSelectedTextRange
Selected text range: Optional(<AXValue 0x600003144060> {value = location:14990 length:20 type = kAXValueCFRangeType})
Selected text: 
Selected text range: Optional(<AXValue 0x600003148360> {value = location:14904 length:0 type = kAXValueCFRangeType})
```

**Comment by intitni** [comment]
Oh I thought there is no such notification.

**Comment by MrMage** [comment]
> Oh I thought there is no such notification. Thanks.

To be honest, I wasn't sure whether it would work, either, so I built the demonstrator to verify 😅

**Comment by intitni** [comment]
Released in 0.17.0
